scripts/smoke/index.js

```diff
@@ -1,33 +1,93 @@
+// @ts-check
+
+import Zip from 'adm-zip';
+import { execa } from 'execa';
+import { polyfill } from '@astropub/webapi';
+import { fileURLToPath } from 'node:url';
+import { promises as fs } from 'node:fs';
+
+polyfill(globalThis, { exclude: 'window document' });
+
+const currentDir = new URL('./', import.meta.url);
+
+/** Returns the URL to the ZIP of the given GitHub project. */
+const getGitHubZipURL = (/** @type {GitHubZipOpts} */ opts) => `https://github.com/${opts.org}/${opts.name}/archive/refs/heads/${opts.branch}.zip`;
+
+/** Returns the awaited ZIP Buffer from the given GitHub project. */
+const fetchGitHubZip = (/** @type {GitHubZipOpts} */ opts) =>
+	fetch(getGitHubZipURL(opts))
+		.then((response) => response.arrayBuffer())
+		.then((arrayBuffer) => Buffer.from(arrayBuffer));
+
+/** Downloads a ZIP from the given GitHub project. */
+const downloadGitHubZip = async (/** @type {GitHubZipOpts} */ opts) => {
+	/** Expected directory when the zip is downloaded. */
+	const expectedDir = new URL(`${opts.name}-${opts.branch}`, currentDir);
+
+	/** Whether the expected directory is already available */
+	const hasExpectedDir = await fs.stat(expectedDir).then(
+		(stats) => stats.isDirectory(),
+		() => false
+	);
+
+	if (!hasExpectedDir) {
+		console.log('🤖', 'Downloading', `${opts.org}/${opts.name}#${opts.branch}`);
+
+		const buffer = await fetchGitHubZip(opts);
+
+		console.log('🤖', 'Extracting', `${opts.org}/${opts.name}#${opts.branch}`);
+
+		new Zip(buffer).extractAllTo(fileURLToPath(currentDir), true);
 
+		console.log('🤖', 'Preparing', `${opts.org}/${opts.name}#${opts.branch}`);
+		await execa('yarn', [], { cwd: fileURLToPath(expectedDir), stdout: 'inherit', stderr: 'inherit' });
+	}
+
+	return expectedDir;
+};
+
+/** Returns all child directories of the given directory. */
+const getDirectories = async (/** @type {URL} */ dir) => {
+	/** @type {URL[]} */
+	const dirs = [];
+
+	for await (const dirent of await fs.opendir(dir)) {
+		if (dirent.isDirectory()) {
+			dirs.push(new URL(dirent.name, dir));
+		}
+	}
+
+	return dirs;
+};
+
+/** Runs all smoke tests. */
+export default async function run() {
 	console.log('');
-	for (const projectToTest of allProjectsToTest) {
-		const filePath = fileURLToPath(projectToTest);
-		console.log('  🤖 Testing', filePath, '\n');
+
+	const directories = await getDirectories(new URL('../../examples/', import.meta.url));
+
+	directories.push(await downloadGitHubZip({ org: 'withastro', name: 'docs', branch: 'main' }), await downloadGitHubZip({ org: 'withastro', name: 'astro.build', branch: 'main' }));
+
+	for (const directory of directories) {
+		console.log('🤖', 'Testing', directory.pathname.split('/').at(-1));
+
 		try {
-			await execa('yarn', ['build'], { cwd: fileURLToPath(projectToTest), stdout: 'inherit', stderr: 'inherit' });
+			await execa('yarn', ['build'], { cwd: fileURLToPath(directory), stdout: 'inherit', stderr: 'inherit' });
```

Comment:
Can we jump directly to `yarn build`? What if these smoke tests have other dependencies that need to be installed?

Reply:
We jump directly to `yarn build` for 2 reasons, but you’ve helped me see another potential issue.

1. External repo tests run `yarn` as part of their initialization process.
2. Internal repo tests use the monorepo version of Astro.
  - `yarn build` will fail if we only run `yarn` from each internal directory.

The issue you have helped me see, however, is that these smoke tests will probably run against their own production versions of astro and _not_ the monorepo version. I would need to sync their package json files and put them in a directory the monorepo would respect. And I might need to be careful, too, because testing them against the monorepo would mean their own dependencies may leak into our `yarn.lock`.

Reply:
This has been resolved in a533e503fc4342a7ac9e2c3d484615616e699bae